go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,321 @@
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	invalidIPAddres  = "192.0.2.254"
+	haLeaderLockName = "ovn-kubernetes-master"
+	ovnkubeDbEp      = "ovnkube-db"
+)
+
+//TODO: Need to come up with proper default values.
+const (
+	leaseDuration = 6 * time.Second
+	renewDeadline = 3 * time.Second
+	retryPeriod   = 2 * time.Second
+)
+
+// OvnHAMasterController is the object holder for managing the HA master
+// cluster
+type OvnHAMasterController struct {
+	kubeClient      kubernetes.Interface
+	cluster         *OvnClusterController
+	ovnController   *ovn.Controller
+	nodeName        string
+	manageDBServers bool
+	isLeader        bool
+	currentLeaderIP string
+}
+
+// NewHAMasterController creates a new HA Master controller
+func NewHAMasterController(kubeClient kubernetes.Interface, cluster *OvnClusterController,
+	nodeName string, manageDBServers bool) *OvnHAMasterController {
+	ovnController := ovn.NewOvnController(kubeClient, cluster.watchFactory)
+	return &OvnHAMasterController{
+		kubeClient:      kubeClient,
+		cluster:         cluster,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		currentLeaderIP: "",
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backmode mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddres, "6641", "6642")
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterNewLeader := func(identity string) {
+		logrus.Infof("The new leader is : " + identity)
+		wasLeader := hacluster.isLeader
+		if hacluster.nodeName == identity {
+			err = hacluster.ConfigureAsActive(identity)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				panic(err.Error())
+			}
+			hacluster.isLeader = true
+		} else {
+			if wasLeader {
```

Comment:
First things first, stop the ovn-northd! Then you can exit.

Reply:
I think this is resolved now.